near-plugins-derive/src/upgradable.rs

```diff
@@ -50,9 +121,64 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
 
             #[#cratename::only(owner)]
             fn up_deploy_code(&mut self) -> near_sdk::Promise {
+                let staging_timestamp = self.up_get_timestamp(__UpgradableStorageKey::StagingTimestamp)
+                    .unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: staging timestamp isn't set"));
+
+                if near_sdk::env::block_timestamp() < staging_timestamp {
+                    near_sdk::env::panic_str(
+                        format!(
+                            "Upgradable: Deploy code too early: staging ends on {}",
+                            staging_timestamp
+                        )
+                        .as_str(),
+                    );
+                }
+
                 near_sdk::Promise::new(near_sdk::env::current_account_id())
                     .deploy_contract(self.up_staged_code().unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: No staged code")))
             }
+
+            #[#cratename::only(owner)]
+            fn up_init_staging_duration(&self, staging_duration: near_sdk::Duration) {
+                near_sdk::require!(self.up_get_duration(__UpgradableStorageKey::StagingDuration).is_none(), "Upgradable: staging duration was already initialized");
+                self.up_set_staging_duration_unchecked(staging_duration);
+            }
+
+            #[#cratename::only(owner)]
+            fn up_stage_update_staging_duration(&self, staging_duration: Option<near_sdk::Duration>) {
+                if let Some(staging_duration) = staging_duration {
+                    let current_staging_duration = self.up_get_duration(__UpgradableStorageKey::StagingDuration)
+                        .unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: staging duration isn't initialized"));
+
+                    let staging_duration_timestamp = near_sdk::env::block_timestamp() + current_staging_duration;
+                    self.up_set_timestamp(__UpgradableStorageKey::StagingTimestamp, staging_duration_timestamp);
```

Comment:
When code is staged, this line shifts the earliest time at which the code can be deployed? What's the reason for that?

Reply:
This is a side effect of using one `Timestamp` for both actions (stage code and stage new duration).
This can be avoided by reverting the requested changes and returning the second `timestamp`, so it will be possible to stage a new duration and code in parallel.

Also as a solution, we could prevent calling this function if the staged code is not empty, but removing the code from the storage after deployment could take extra gas so there is a risk to get out of gas problem.

Reply:
> we could prevent calling this function if the staged code is not empty

This was part of the proposed idea, see [this comment](https://github.com/aurora-is-near/near-plugins/pull/44#discussion_r1090401902).

Reply:
Reverted in c1654aea3df463524579d02fa8a091b8c59fb017